proposals/list-patterns.md

```diff
@@ -0,0 +1,95 @@
+
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is [1, 2, 3]` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+### Fixed-length list patterns
+
+The syntax will be modified to include a *list_pattern* defined as below:
+
+```antlr
+primary_pattern
+	: list_pattern
+list_pattern
+	: '[' (pattern (',' pattern)* ','?)? ']'
+```
+
+**Pattern compatibility:** A *list_pattern* is compatible with any type that conforms to the ***range indexer pattern***:
+
+1. Has an accessible property getter that returns an `int` and has the name `Length` or `Count`
+2. Has an accessible indexer with a single `int` parameter
+3. Has an accessible `Slice` method that takes two `int` parameters (for slice patterns)
+
+This rule includes `T[]`,  `string`,  `Span<T>`, `ImmutableArray<T>` and more.
+
+**Subsumption checking:**  This construct will be built entirely on top of the existing DAG nodes. The subsumption is checked just like recursive patterns with `ITuple` - corresponding subpatterns are matched by position plus an additional node for testing length.
+
+**Lowering:** A pattern of the form `expr is [1, 2, 3]` is equivalent to the following code:
+```cs
+expr.Length is 3
```

Comment:
Does this need to include a null check?

```cs
expr is not null
&& expr.Lenght is 3
&& expr[0] is 1
&& expr[1] is 2
&& expr[2] is 3
```

Alternatively, could it use a property pattern

```cs
expr is { Lenght: 3 }
&& expr[0] is 1
&& expr[1] is 2
&& expr[2] is 3
```

Reply:
yes.  :)